### DISCUSSION TITLE
Step Doc

### DISCUSSION THREAD

**Discussion by Darius4000** [Azure/ipam#43] [General]
GA, 
I am planning to deploy Azure IPAM with the solution as an Azure Function (server Less),
is there a step-by-step document to follow for Azure function deployment?
I see a document included with Azure IPAM in GitHub, but it’s general.

Thanks

**Comment by DCMattyG** [comment]
Yes @Darius4000, you can reference the "Full Deployment" section in our documentation here:

https://azure.github.io/ipam/#/deployment/README?id=full-deployment

Scroll down just a little, and you'll see the section titled "Deploy IPAM solution as an Azure Function".

Hope that helps, let me know if you have any issues!

**Reply by Darius4000** [reply]
Are you mentioning the script below from the document?

/deploy.ps1 `
  -Location "westus3" `
  -AsFunction

**Comment by DCMattyG** [comment]
Yes indeed! By adding the "-AsFunction" switch, Azure IPAM will automatically deploy only the Engine components into an Azure Function instead of a multi-container App Service.

Were you looking for something else, perhaps a walk-through of how to manually configure these components in the Azure Portal and link them together?

**Reply by Darius4000** [reply]
Yes, steps a walk-through of how to manually configure these components in the Azure Portal and link them together, plus requirements for only Azure function solution method deployment.

**Reply by DCMattyG** [reply]
That's a fair ask. Currently we are working on some docs describing just this for deployments such as Functions, AKS, and more.

For now, I can tell you that by using the "-AsFunction", everything is the same with the exception of:

1) The storage account is now used for Function storage instead of NGINX config storage
2) The UI App Registration is no longer required/created
3) Engine container is deployed into an Azure Function instead of an App Service

Is there a reason, for the time being, that using the deployment script isn't an option for you and you want to build everything manually?

**Reply by Darius4000** [reply]
Hi, Is there a revert option for this script in case we decide to remove it completely?

**Reply by Darius4000** [reply]
Hi, I wonder if you haven't seen my question, Is there a revert option for this script in case we decide to remove it completely? As I mentioned, planning to deploy it on Prod since there is no Lab in our environment. Thanks

**Reply by DCMattyG** [reply]
Hi @Darius4000, apologies for missing this. Currently we don't have a "uninstall" script, that said it should be a very simple process:

1) Delete the Resource Group, that will get rid of all of the Azure resources
2) Delete the 2x App Registrations that were created
3) Delete the stale "Reader" permission from the Root Management Group (now that the SP is gone)

That should be it. We'll work on perhaps the Deploy script generating a removal script during it's final steps. Not a bad idea.

Hope that helps for now!

**Reply by Darius4000** [reply]
GM Matt,
Question, deploying IPAM solution as an Azure Function (-AsFunction), what components will be installed? I have asked to calculate the cost and I want to make sure they have it right. Thanks

**Reply by DCMattyG** [reply]
You bet @Darius4000!

Everything that is deployed is documented here:
https://azure.github.io/ipam/#/README?id=ipam-infrastructure

You'll see there is a note on which pieces are different for Full vs. Function only.

Is that what you are looking for, or did you need some more specifics on those items?

**Reply by DCMattyG** [reply]
-The Function App is set to an EP1 SKU by our deployment scripts, but this can be changed to fit your needs.
-Cosmos is created with Autoscale (fairly low usage)
-Storage Account is a V2 type
-KeyVault is a Standard SKU

**Reply by Darius4000** [reply]
Great, That's what I was looking for, Cosmos DB, Storage account, and key vault.  3 components.

**Reply by Darius4000** [reply]
GM Matt, 
do you have an instruction on how to use override default resource names with custom resource names?
Could you guide me since there is no instruction as part of the IPAM document? Thanks!
[ResourceNames.docx](https://github.com/Azure/ipam/files/10004281/ResourceNames.docx)
[

**Reply by DCMattyG** [reply]
Hi @Darius4000, there is a note at the bottom of the Deployment section talking about how to formulate these names. For quick reference, it links here: https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules

TL:DR - The names must adhere to the standard naming rules for each Azure resource (e.g. if the name validates in the Azure Portal, it will validate in the IPAM Deployment, we use the same Regex).

I hope that helps....the idea is we're not doing anything special with the names in IPAM, we just use the exact same validation rules that Azure does.

**Reply by Darius4000** [reply]
I saw that; I was referring to an example script, A-  should I save the Resource name sample I sent you to a script, then run it? Or B- add my custom names to IPAM deploy Script and run it?

**Reply by DCMattyG** [reply]
Hey @Darius4000, hopefully I'm understanding correctly. Your question is around the best way to execute this, correct?

The simplest way is to just copy & paste the variable assignment (from your document) into a PowerShell window. PowerShell support multi-line pasting, so that should be sufficient to set the variable (HashTable) for the current PS session. After that, as the variable $ResourceNames has been created, you can execute the deployment script by passing in the HashTable as an argument.

Just like you have in the above attached doc:

./deploy.ps1 -Location "westus3" -ResourceNames $ResourceNames

I truly hope that helps, and if I have misunderstood, just let me know. Thanks!

**Reply by DCMattyG** [reply]
Alternatively, you can create another PS1 script which sets the ResourceNames variable, and then subsequently executes the Deployment script. It's really up to you how you'd prefer to do it (there's multiple ways).

**Reply by Darius4000** [reply]
Thanks!

**Reply by Darius4000** [reply]
GM Matt,
During the full deployment in the commercial gov cloud, the storage account did not deploy, but worked when I deployed as a function method, is it necessary to have it, and what role is playing?

**Reply by DCMattyG** [reply]
Good morning @Darius4000. Originally, when doing a "full" deployment, the function of the Storage Account was to store the NGINX configuration. Recently I've made some updates where the NGINX config is now baked into the NGINX container image itself, so the Storage Account in this deployment type is no longer required.

For a "function" deployment, a Storage Account is still require to be attached to the Function App for metadata, logging, etc.

Hope that helps to clarify everything for you!

**Reply by Darius4000** [reply]
GM Matt,
I have logged in to Azure IPAM console I set myself as admin, but I am not able to discover subnet, vnet, or any other options, I tried to add space, but I am not able to enter the name or create since it's grayed out, my team member also tried the same thing it was the same, I also tried it in my personal lab, the same issue, what could be the problem?

**Reply by DCMattyG** [reply]
Hi @Darius4000, I'm assuming you added yourself as an admin via the UI and not via API, correct? If so, did you click the "save" button that appears in the upper right when names are added/subtracted from the admin list? When you navigate back to the admins page, is your name & details still visible there?

Additionally, if you'd like to troubleshoot this more 1:1 with me, don't hesitate to reach out to me at Matthew.Garrett@microsoft.com.

**Reply by Darius4000** [reply]
Thanks!

**Reply by Darius4000** [reply]
Hi, Matt
Thanks for your help. Permission is fixed and IPAM is working now. I

**Reply by Darius4000** [reply]
GM Matt, I was wondering if the new release for Azure IPAM you mentioned is ready. Thanks.

**Reply by DCMattyG** [reply]
Hi @Darius4000, apologies for the delay. I'll be releasing the new code this coming weekend. Hope you enjoy!
